### DISCUSSION TITLE
How to get caller name from a module (in strict mode)

### DISCUSSION THREAD

**Discussion by maciej-wakula** [nodejs/help#4785] [Q&A]
**Goal:** Obtain caller name for instrumentation purpose from utility/framework.

**Expected example:**

```
console.log('anonymous call');
function f1() {
  console.log('I am f1')
  f2();
}
const f2 = async function() {
  await console.log('I am f2')
  f3();
}
const f3 = () => {
  console.log('I am f3')
}
```

It would be nice to print something like

```
"undefined (file:///index.js)": "I am anonymous"
"f1": "I am f1"
"f2": "I am f2"
"f3": "I am f3"
```

Problems:
- In strict mode call stack frame is not accessible. Every node module runs in strict mode.
- f2 and f3 are anonymous so it is well justified that they contain no name.
- I am not sure if hoisting would impact it in any way
- Since we can access call stack anyway by throwing an error and read its stack - I don't see how access restriction to caller name would increase security

Commonly used solutions:
- use non-strict mode if possible
- throw an error, read its stack, find who might be the caller
  - encounter funny issues when some other instrumentation wrapped your function
  - some are even replacing V8 functions with custom made before creating new stack
- rewrite the code and replace every call with a wrapper knowing who is being called
  - but then the wrapper (decorator function) becomes actually new re-caller

Anyone any ideas how we could improve this to get the name of actual caller of our function without hacks like new error stack?

**Comment by cjihrig** [comment]
I think your options are pretty much limited to:

1. Parsing an error stack. I know you mentioned not wanting to do this, and it is fragile.
2. Using [V8's stack trace API](https://v8.dev/docs/stack-trace-api). This is what Node does internally a lot, but it's also not portable as far as I know.
3. Use C++ code. Node also uses this approach internally. The obvious downside is needing to use a compiled addon.

**Reply by maciej-wakula** [reply]
(2.) is what some modules are using - throwing an error and then "capturing" methods from the error object, often replacing them.
It would be nice if a module would be able to know its direct caller name. And, since this can be obtained from call stack anyway, lack of caller provides absolutely no extra security in my opinion. For completeness [caller  method on mozilla.org](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/caller) with its clumsy justification.
